[WIP] Introduce SourcekitVariant

norio-nomura opened the pull request:
Add `SourceKitVariant` that replacing `SourceKitRepresentable`.
- Replace complicated `SourceKitRepresentable`'s  operations to simple operations
- No performance penalty
  - Same on whole data parsing
  - Better on partial data parsing

``` swift
func walk1(dictionay: [String: SourceKitRepresentable]) {
    if let _ = dictionay[SwiftDocKey.name.rawValue] as? String {
    }
    if let _ = dictionay[SwiftDocKey.kind.rawValue] as? String {
    }
    if let _ = (dictionay[SwiftDocKey.offset.rawValue] as? Int64).map({Int($0)}) {
    }
    if let _ = (dictionay[SwiftDocKey.length.rawValue] as? Int64).map({Int($0)}) {
    }
    guard let substructures = dictionay[SwiftDocKey.substructure.rawValue] as? [SourceKitRepresentable] else { return }
    for substructure in substructures {
        if let substructure = substructure as? [String: SourceKitRepresentable] {
            walk1(dictionay: substructure)
        }
    }
}
```

will be changed to:

``` swift
func walk2(variant: SourceKitVariant) {
    if let _ = variant[.name]?.string {
    }
    if let _ = variant[.kind]?.string {
    }
    if let _ = variant[.offset]?.int {
    }
    if let _ = variant[.length]?.int {
    }
    guard let substructures = variant[.substructure]?.array else { return }
    for substructure in substructures {
        walk2(variant: substructure)
    }
}
```

jpsim commented:
This looks pretty great! Can we wait to get this in after SPM and Linux support though? That's #267 and #268.

There's no need to keep the old versions in there, you can delete that old code 😄. I know you probably just kept it to measure the perf difference, but I'm convinced!

norio-nomura commented:
I added FindAvailables performance tests.
- testFindAvailablesWithDictionary:
> [SourceKittenFrameworkTests.PerformanceTests testFindAvailablesWithDictionary]' measured [Time, seconds] average: 0.001, relative standard deviation: 25.468%, values: [0.001627, 0.001280, 0.001151, 0.000837, 0.000833, 0.001112, 0.000839, 0.000832, 0.000825, 0.000834], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100

- testFindAvailablesWithVariant:
> [SourceKittenFrameworkTests.PerformanceTests testFindAvailablesWithVariant]' measured [Time, seconds] average: 0.001, relative standard deviation: 209.061%, values: [0.005857, 0.000391, 0.000288, 0.000213, 0.000252, 0.000217, 0.000210, 0.000210, 0.000211, 0.000209], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100

norio-nomura commented:
rebased.

norio-nomura commented:
I added "RequestEditorOpen" tests.
- testRequestEditorOpenWithDictionary:
> '-[SourceKittenFrameworkTests.VariantPerformanceTests testRequestEditorOpenWithDictionary]' measured [Time, seconds] average: 0.021, relative standard deviation: 8.027%, values: [0.024927, 0.020399, 0.020207, 0.020037, 0.019598, 0.021164, 0.019854, 0.020056, 0.023145, 0.019551], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100

- testRequestEditorOpenWithVariant:
> '-[SourceKittenFrameworkTests.VariantPerformanceTests testRequestEditorOpenWithVariant]' measured [Time, seconds] average: 0.011, relative standard deviation: 8.340%, values: [0.012964, 0.010267, 0.009695, 0.010078, 0.010301, 0.010115, 0.010215, 0.010730, 0.011066, 0.010083], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100

`SourceKitVariant` converts SourceKit values into Swift values lazily. So, `failableSend2()` is faster than `failableSend()`.
The results in https://github.com/jpsim/SourceKitten/pull/274#issuecomment-254697528 shows that the first access is slow on `SourceKitVariant`. That is trade-off to them.

I expect that overall performance will increase by `SourceKitVariant` because of parsing operation is 4~5 time faster than `SourceKitRepresentable`.

norio-nomura commented on Source/SourceKittenFramework/SourceKitVariant.swift:
```diff
@@ -0,0 +1,345 @@
+    var nameLength: Int? { return self[.nameLength]?.int }
+    var nameOffset: Int? { return self[.nameOffset]?.int }
+    var offset: Int? { return self[.offset]?.int }
+    var substructure: [SourceKitVariant]? { return self[.substructure]?.array }
```
I'm considering that `substructure` would returns non optional.
```diff
diff --git a/Source/SourceKittenFramework/SourceKitVariant.swift b/Source/SourceKittenFramework/SourceKitVariant.swift
index e8e5ee5..16641e1 100644
--- a/Source/SourceKittenFramework/SourceKitVariant.swift
+++ b/Source/SourceKittenFramework/SourceKitVariant.swift
@@ -332,10 +332,10 @@ extension SourceKitVariant {
     var nameLength: Int? { return self[.nameLength]?.int }
     var nameOffset: Int? { return self[.nameOffset]?.int }
     var offset: Int? { return self[.offset]?.int }
-    var substructure: [SourceKitVariant]? { return self[.substructure]?.array }
-    var syntaxMap: [SourceKitVariant]? { return self[.syntaxMap]?.array }
+    var substructure: [SourceKitVariant] { return self[.substructure]?.array ?? [] }
+    var syntaxMap: [SourceKitVariant] { return self[.syntaxMap]?.array ?? [] }
     var typeName: String? { return self[.typeName]?.string }
-    var inheritedtypes: [SourceKitVariant]? { return self[.inheritedtypes]?.array }
+    var inheritedtypes: [SourceKitVariant] { return self[.inheritedtypes]?.array ?? [] }
 }
 
 // MARK: - Custom
```
With current implementation:
```swift
variant.substructure.map { $0 in // $0 is [SourceKitVariant] }
variant.substructure?.map { $0 in // $0 is SourceKitVariant }
```
will be changed to:
```swift
variant.substructure.map { $0 in // $0 is SourceKitVariant }
variant.substructure?.map { $0 in } // invalid
```
Checking existence of `key.substructure` can be achieved by another way: `variant[.substructure] != nil`

I think those properties are convenience accessor, and should be more convenient.

norio-nomura replied:
> should be more convenient.

I changed my thought.
`let` binding in middle of `if`/`guard` condition is useful, though `?` is required.

jpsim commented on Source/SourceKittenFramework/SourceKitVariant.swift:
```diff
@@ -0,0 +1,447 @@
+    public var offset: Int? { return self[.offset]?.int }
+    /// Substructure ([SourceKitVariant]).
+    public var substructure: [SourceKitVariant]? { return self[.substructure]?.array }
+    /// Syntax map ([SourceKitVariant]).
```
Maybe this could return `Data` or `SyntaxMap` instead?

jpsim commented on Source/SourceKittenFramework/SourceKitVariant.swift:
```diff
@@ -0,0 +1,447 @@
+    public var inheritedtypes: [SourceKitVariant]? { return self[.inheritedtypes]?.array }
+}
+
+// MARK: - Accessors for SwiftDocKey
```
Ideally we'd just remove `SwiftDocKey`. Or at least just make it a private implementation detail.

jpsim commented on Source/SourceKittenFramework/SourceKitVariant.swift:
```diff
@@ -0,0 +1,447 @@
+            switch sourcekitd_variant_get_type(sourcekitObject) {
+            case SOURCEKITD_VARIANT_TYPE_ARRAY:
+                var array = [SourceKitVariant]()
+                _ = sourcekitd_variant_array_apply(sourcekitObject) { index, value in
```
These might need to be adapted to work on Linux once #268 lands.

norio-nomura commented:
rebased. Now this runs on Linux.

norio-nomura commented on Tests/SourceKittenFrameworkTests/VariantPerformanceTests.swift:
```diff
@@ -0,0 +1,146 @@
+
+let largestSwiftFile = File(path: largestSwiftFileInRepoURL.path)!
+
+let thisFile = URL(fileURLWithPath: #file).path
```
Yeah! 🙈

norio-nomura commented:
I created `UID` that represents `sourcekitd_uid_t` in Swift, corrected `UIdent` definitions from SourceKit and added them to `UID` as static properties.
I think those properties would be replacement of previous enums such as `SwiftDocKey`.

By those changes, parsing operations become about 2 time faster than first implementation:
- testFindAvailablesWithVariant:

> -[SourceKittenFrameworkTests.VariantPerformanceTests testFindAvailablesWithVariant]' started.
/Users/norio/github/swift-dev/SourceKitten/Tests/SourceKittenFrameworkTests/VariantPerformanceTests.swift:128: Test Case '-[SourceKittenFrameworkTests.VariantPerformanceTests testFindAvailablesWithVariant]' measured [Time, seconds] average: 0.001, relative standard deviation: 252.375%, values: [0.004816, 0.000104, 0.000089, 0.000088, 0.000087, 0.000087, 0.000087, 0.000087, 0.000087, 0.000087], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100

norio-nomura commented:
@jpsim I pushed some commits that replacing some implementation with using `SourceKitVariant`.
Those will be some examples for real usage.

norio-nomura commented on Source/SourceKittenFramework/SourceKitVariant.swift:
```diff
@@ -0,0 +1,674 @@
+        set { self[Custom.docParameters] = SourceKitVariant(newValue) }
+    }
+
+    /// Doc result discussion ([SourceKitVariant]).
```
These keys comes from SwiftDocKey.swift.
I noticed that documentation comments of them are incorrect.
I copied and modified them to here, but originals are not modified.
https://github.com/jpsim/SourceKitten/blob/master/Source/SourceKittenFramework/SwiftDocKey.swift#L50-L87

norio-nomura commented:
Codes which became unnecessary still remains, but it has generally been rewritten.
Also, I want to do additional things,
- [ ] Clean up unnecessary codes.
- [ ] Implement Value type that representing `sourcekitd_request_*` that using `UID`.
- [ ] Converting byte location/length to utf16 range as properties of `SourceKitVariant`. (Another PR?)

norio-nomura commented:
@jpsim Do you think loading multiple SourceKit should be considered on this or future PR? ( https://github.com/realm/jazzy/issues/663#issuecomment-263037402)

jpsim commented:
Let's not add more to this than we have to!

![image](https://cloud.githubusercontent.com/assets/474794/20700841/8b881f60-b5c4-11e6-8dcb-a423a524d721.png)

norio-nomura commented:
Ok, I won't.

norio-nomura commented:
Added `SourceKitObject` that represents `sourcekitd_object_t` in Swift.

jpsim commented:
@norio-nomura what's the state of this? What's next here?

norio-nomura commented:
@jpsim I think it needs to try using this in SwiftLint and confirm this would work well.
After that, I would open another PR implementing "Converting byte location/length to utf16 range as properties of `SourceKitVariant`" that would be expected to deprecate byte offset operations of `String`.

I'm not yet sure if this concept is right. Maybe it would needs to try using utf16 range properties in SwiftLint.
Sorry for slow of progress.

jpsim commented:
No problem at all! I was just wondering if you were waiting on me for something or this was still in progress.

norio-nomura commented:
As far as I see the change in https://github.com/jpsim/SourceKitten/pull/274/commits/50caf4c296da37057f31f550ba0bf2afa08c89b3, the way to define all known UIDs depends strongly on Swift's version, so it seems undesirable. 🤔

norio-nomura commented:
Sorry to leave it long. 🙇 
I do not feel like I can finish this anymore, and it is not my intention to prevent new efforts like #428, so I will close it.
I think that it was not good that the change around `UID` became too big.
